src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1024 @@
+					res[bit_idx][tray_idx] = GetSilh(pointsets[T], start_trays[bit_idx][tray_idx], tmp_cluster);
+				}
+				// end Silhoutte 
+
+
+				// begin capacitated k-means clustering
+
+				for (int i = 1; i < 6; i++) {
+
+						int rows = GetRows(BITCNT[i]), cols = BITCNT[i] / rows;
+						
+
+						int num_trays = (num_flops + (BITCNT[i] - 1)) / BITCNT[i];
+
+						int opt_idx = 0;
+						float opt_val = -1;
+						for (int j = 0; j < 5; j++) {
+							if (res[i][j] > opt_val) {
+								opt_val = res[i][j];
+								opt_idx = j;
+							}
+						}
+
+						trays[i] = start_trays[i][opt_idx];
+
+
+
+						for (int j = 0; j < num_trays; j++) trays[i][j].slots = GetSlots(trays[i][j].pt, rows, cols);
+
+						// clusters[i] = (tray that flop i belongs to, slot that flop i belongs to)
+						float delta = 0;
+						for (int j = 0; j < 35; j++) {
+		            		clusters[i] = MinCostFlow(pointsets[T], trays[i], BITCNT[i]);
+							delta = MBFF::RunLP(pointsets[T], trays[i], clusters[i], BITCNT[i]);
+							for (int k = 0; k < num_trays; k++) {
+								trays[i][k].slots = GetSlots(trays[i][k].pt, rows, cols);
+							}
+
+		            		if (delta < 0.5) break;
+						}
+
+						 clusters[i] = MinCostFlow(pointsets[T], trays[i], BITCNT[i]);
+
+						for (int j = 0; j < num_trays; j++) {
+							trays[i][j].slots = GetSlots(trays[i][j].pt, rows, cols);
+						}
+				}
+				// end capacitated k-means clustering
+
+
+				// remap the smaller pointset to the larger one
+				map<int, int> small_to_large;
+				for (int i = 0; i < num_flops; i++) {
+					for (int j = 0; j < (int)flops.size(); j++) {
+						if (pointsets[T][i].pt.x == flops[j].pt.x && pointsets[T][i].pt.y == flops[j].pt.y) {
+							small_to_large[i] = j;
+							break;
+						}
+					}
+				}
+
+				for (int i = 0; i < 6; i++) {
+					for (int j = 0; j < (int)trays[i].size(); j++) {
+						for (int k = 0; k < 70; k++) {
+							trays[i][j].cand[k] = small_to_large[trays[i][j].cand[k]];
+						}
+					}
+				}
+
+				// append trays 
+				for (int i = 0; i < 6; i++) {
+					for (int j = 0; j < (int)trays[i].size(); j++) {
+						all_trays[i].push_back(trays[i][j]);
+					}
+				}
+    	}
+
+    	MBFF::RunILP(flops, paths, all_trays, ALPHA, BETA);
+	}
+
+
+	// ctor
+	MBFF::MBFF(int num_flops, int num_paths, vector<float> x, vector<float> y, vector<pair<int, int> > paths, int threads) {
+		FLOPS.resize(num_flops);
+		for (int i = 0; i < num_flops; i++) {
+			Point new_pt; 
+			new_pt.x = x[i] + 70.0, new_pt.y = y[i] + 70.0;
+
+			Flop new_flop; 
+			new_flop.pt = new_pt, new_flop.idx = i;
+			FLOPS[i] = new_flop;
+		}
+
+		PATHS.resize(num_paths);
+		for (int i = 0; i < num_paths; i++) {
+			Path new_path;
+			new_path.a = paths[i].first, new_path.b = paths[i].second;
+			PATHS[i] = new_path;
+		}
+
+		NUM_THREADS = threads;
+	}
+
+	// dector
+	MBFF::~MBFF() {
+		FLOPS.clear();
+		PATHS.clear();
+		NUM_THREADS = 0;
+	}
```

Comment:
Not necessary. The default destructor will handle this.

Please replace with MBFF::~MBFF() = default;